File: Data Structures & Algorithms/mergeSort/submission-1.js
/** Pair class to store key-value pairs */
// class Pair {
//   /**
//    * @param {number} key The key to be stored in the pair
//    * @param {string} value The value to be stored in the pair
//    */
//   constructor(key, value) {
//       this.key = key;
//       this.value = value;
//   }
// }
class Solution {
    /**
     * @param {Pair[]} pairs
     * @returns {Pair[]}
     */
    mergeSort(pairs) {
        return this.mergeSortHelper(pairs, 0, pairs.length - 1)
    }

    mergeSortHelper(pairs, left, right) {
        if (left < right) {
            const middle = Math.floor((left + right) / 2)

            this.mergeSortHelper(pairs, left, middle)
            this.mergeSortHelper(pairs, middle + 1, right)
            this.merge(pairs, left, middle, right)
        }

        return pairs
    }

    merge(pairs, left, middle, right) {
        const length1 = middle - left + 1
        const length2 = right - middle

        let leftSubArr = new Array(length1)
        let rightSubArr = new Array(length2)

        for (let i = 0; i < length1; i++) {
            leftSubArr[i] = pairs[left + i]
        }

        for (let j = 0; j < length2; j++) {
            rightSubArr[j] = pairs[middle + 1 + j]
        }

        let i = 0
        let j = 0
        let k = left

        while (i < length1 && j < length2) {
            if (leftSubArr[i].key <= rightSubArr[j].key) {
                pairs[k] = leftSubArr[i]
                i++
            } else {
                pairs[k] = rightSubArr[j]
                j++
            }
            k++
        }

        while (i < length1) {
            pairs[k] = leftSubArr[i]
            i++
            k++
        }

        while (j < length2) {
            pairs[k] = rightSubArr[j]
            j++
            k++
        }
    }
}

File: Data Structures & Algorithms/sort-an-array/submission-1.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def merge(arr, L, R, M):
            left, right, i = 0, 0, L
            leftArr = arr[L:M+1]
            rightArr = arr[M+1:R+1]

            while left < len(leftArr) and right < len(rightArr):
                if leftArr[left] <= rightArr[right]:
                    arr[i] = leftArr[left]
                    i += 1
                    left += 1
                else:
                    arr[i] = rightArr[right]
                    i += 1
                    right += 1
            
            while left < len(leftArr):
                arr[i] = leftArr[left]
                i += 1
                left += 1
            
            while right < len(rightArr):
                arr[i] = rightArr[right]
                i += 1
                right += 1


        def mergeSort(arr, L, R):
            if L >= R:
                return arr

            mid = (R + L) // 2
            mergeSort(arr, L, mid)
            mergeSort(arr, mid + 1, R)
            merge(arr, L, R, mid)
            return arr

        return mergeSort(nums, 0, len(nums) - 1)